proposals/0063-feature-flag.md

```diff
@@ -0,0 +1,137 @@
+
+- Prometheus >= 2.24.0
+- Mimir >= 2.0.0
+- Cortex >= 1.11.0
+- Thanos >= 0.18.0
+
+If the criteria are met, Grafana chooses more efficient label endpoints (`/api/v1/labels`, `/api/v1/label/<name>/values` with `match[]`). Otherwise, it falls back to the less efficient `/api/v1/series` for label queries.
+
+Key limitations of this approach:
+
+1. Configuration errors (wrong type or version) can lead to incompatible or missing features.
+2. Backend upgrades alone do not enable new features in clients—client logic must also be updated and released.
+3. New Prometheus-compatible backends require explicit code changes in Grafana, slowing adoption.
+4. Type and version checks are coarse; they do not reflect actual enabled features, which may depend on flags or configuration.
+
+Alternatives already exist in some downstream projects and demonstrate the need for such kind of APIs. However, the current approach is based on extending the [`buildinfo` endpoint](https://prometheus.io/docs/prometheus/latest/querying/api/#build-information) with a [`features` field](https://github.com/grafana/mimir/blob/9fccbacdabdd236cb7ff97cf154643b409078178/pkg/util/version/info_handler.go#L11-L30), which is very vendor specific. Grafana already uses this approach for some [alertmanager features](https://github.com/grafana/grafana/blob/8863ed9d6f8395808196b5d81d436fb637a43d37/public/app/features/alerting/unified/api/buildInfo.ts#L137-L145).
+
+## Goals
+
+- Provide a machine-readable API to report enabled features.
+- Ensure the solution is lightweight to encourage broad adoption in the ecosystem.
+- Cover a comprehensive and relevant subset of Prometheus features.
+- Design the API to be extensible, allowing third-party projects to declare their own features.
+- Allow for potential future inclusion of Alertmanager, even though it is currently out of scope.
+
+### Audience
+
+The intended audience for this proposal includes:
+
+- Developers creating software that exposes the Prometheus API
+- Consumers of the Prometheus API
+
+## Non-Goals
+
+Implementing a unified feature gate in the code is out of scope
+
+## How
+
+The `/api/v1/features` endpoint returns a JSON object with top-level categories inspired by Prometheus package organization. Each category key contains a map of unique feature names (strings) to `true`/`false` booleans indicating whether the feature is enabled.
+
+Initial categories:
+
+- `api` - API endpoint features and capabilities
+- `remote_write_receiver` - Remote write receiver features
+- `remote_write_sender` - Remote write sender features
+- `scrape` - Scraping capabilities
+- `tsdb` - Time series database features
+- `rule` - Rule evaluation features
+- `ui` - Web UI capabilities
+- `promql` - PromQL language features (syntax, modifiers, operators)
+- `promql_functions` - Individual PromQL functions
+
+Example response:
+
+```json
+{
+  "status": "success",
+  "data": {
```

Comment:
How would the requirement "Allow for potential future inclusion of Alertmanager, even though it is currently out of scope." look like in this API?

I mean I can imagine alertmanager simply returning a similar JSON like this, so I'm not sure what inclusion means?

Reply:
I removed the requirement. I expect that if Alertmanager introduces this they will follow the same principles naturally.